File: trees_and_graphs/keys_and_rooms.java
class Solution
{
    Set<Integer> seen = new HashSet<>();

    public void dfsRecursive(int node, List<List<Integer>> rooms)
    {
        for (int neighbor : rooms.get(node))
        {
            if (!seen.contains(neighbor))
            {
                seen.add(neighbor);
                dfsRecursive(neighbor, room);
            }
        }
    }

    public boolean canVisitAllRooms(List<List<Integer>> rooms)
    {
        seen.add(0);
        dfsRecursive(0, rooms);
        return seen.size() == rooms.size();
    }
}

File: trees_and_graphs/keys_and_rooms.py
class Solution:
    def canVisitAllRooms(rooms: list[list[int]]) -> bool:
        def dfsRecursive(node):
            for neighbor in rooms[node]:
                if neighbor not in seen:
                    seen.add(neighbor)
                    dfsRecursive(neighbor)
        
        seen = {0}
        dfsRecursive(0)
        return len(seen) == len(rooms)